### PR TITLE
GM

### PR DESCRIPTION
**Pull request by mayuanchenma**

### CODE REVIEW

File: tools/tpm2_certify.c [hunk 1]
```diff
@@ -296,6 +255,14 @@ static bool on_option(char key, char *value) {
             return false;
         }
         break;
+    case 'e':
```
Comment (idesai) [inline tools/tpm2_certify.c:258]:
```suggestion
    case 2:
```

File: tools/tpm2_certify.c [hunk 2]
```diff
@@ -336,6 +303,7 @@ static bool tpm2_tool_onstart(tpm2_options **opts) {
       { "certifiedkey-auth",    required_argument, NULL, 'p' },
       { "signingkey-auth",      required_argument, NULL, 'P' },
       { "hash-algorithm",       required_argument, NULL, 'g' },
+      { "scheme",               required_argument, NULL, 'e' },
```
Comment (idesai) [inline tools/tpm2_certify.c:306]:
```suggestion
      { "scheme",               required_argument, NULL, 2 },
```

File: tools/tpm2_certify.c [hunk 3]
```diff
@@ -344,7 +312,7 @@ static bool tpm2_tool_onstart(tpm2_options **opts) {
       { "session",              required_argument, NULL, 'S' },
     };
 
-    *opts = tpm2_options_new("P:p:g:o:s:c:C:f:S:", ARRAY_LEN(topts), topts,
+    *opts = tpm2_options_new("P:p:g:e:o:s:c:C:f:S:", ARRAY_LEN(topts), topts,
```
Comment (idesai) [inline tools/tpm2_certify.c:315]:
```suggestion
    *opts = tpm2_options_new("P:p:g:o:s:c:C:f:S:", ARRAY_LEN(topts), topts,
```

File: tools/tpm2_createek.c [hunk 4]
```diff
@@ -194,8 +196,17 @@ static tool_rc set_ek_template(ESYS_CONTEXT *ectx, TPM2B_PUBLIC *input_public) {
...
         break;
     case TPM2_ALG_ECC:
-        template_nv_index = ECC_EK_TEMPLATE_NV_INDEX;
-        nonce_nv_index = ECC_EK_NONCE_NV_INDEX;
+        if (input_public->publicArea.parameters.eccDetail.curveID == TPM2_ECC_NIST_P256) {
+            template_nv_index = ECC_EK_TEMPLATE_NV_INDEX;
+            nonce_nv_index = ECC_EK_NONCE_NV_INDEX;
+        } else if (input_public->publicArea.parameters.eccDetail.curveID == TPM2_ECC_SM2_P256) {
+            template_nv_index = ECC_SM2_EK_TEMPLATE_NV_INDEX;
+            // EK Nonces SHALL NOT be Populated in any NV Index in the High Range.
```
Comment (idesai) [inline tools/tpm2_createek.c:204]: is this specific to SM2?

File: tools/tpm2_getcommandauditdigest.c [hunk 5]
```diff
@@ -103,10 +113,11 @@ static bool tpm2_tool_onstart(tpm2_options **opts) {
         { "signature",      required_argument, NULL, 's' },
         { "message",        required_argument, NULL, 'm' },
         { "format",         required_argument, NULL, 'f' },
-        { "hash-algorithm", required_argument, NULL, 'g' }
+        { "hash-algorithm", required_argument, NULL, 'g' },
+        { "scheme",         required_argument, NULL, 'e' },
```
Comment (idesai) [inline tools/tpm2_getcommandauditdigest.c:117]:
```suggestion
        { "scheme",         required_argument, NULL, 0 },
```

File: tools/tpm2_getsessionauditdigest.c [hunk 6]
```diff
@@ -93,6 +95,14 @@ static bool on_option(char key, char *value) {
             return false;
         }
         break;
+    case 'e':
```
Comment (idesai) [inline tools/tpm2_getsessionauditdigest.c:98]:
```suggestion
    case 0:
```

File: tools/tpm2_quote.c [hunk 7]
```diff
@@ -227,6 +229,14 @@ static bool on_option(char key, char *value) {
             return false;
         }
         break;
+    case 'e':
```
Comment (idesai) [inline tools/tpm2_quote.c:232]:
```suggestion
    case 1:
```

File: tools/tpm2_quote.c [hunk 8]
```diff
@@ -248,10 +258,11 @@ static bool tpm2_tool_onstart(tpm2_options **opts) {
         { "pcrs_format",    required_argument, NULL, 'F' },
         { "format",         required_argument, NULL, 'f' },
         { "hash-algorithm", required_argument, NULL, 'g' },
+        { "scheme",         required_argument, NULL, 'e' },
```
Comment (idesai) [inline tools/tpm2_quote.c:261]:
```suggestion
        { "scheme",         required_argument, NULL, 1 },
```

File: tools/tpm2_quote.c [hunk 9]
```diff
@@ -248,10 +258,11 @@ static bool tpm2_tool_onstart(tpm2_options **opts) {
...
         { "cphash",         required_argument, NULL,  0  }
     };
 
-    *opts = tpm2_options_new("c:p:l:q:s:m:o:F:f:g:", ARRAY_LEN(topts), topts,
+    *opts = tpm2_options_new("c:p:l:q:s:m:o:F:f:g:e:", ARRAY_LEN(topts), topts,
```
Comment (idesai) [inline tools/tpm2_quote.c:265]:
```suggestion
    *opts = tpm2_options_new("c:p:l:q:s:m:o:F:f:g:", ARRAY_LEN(topts), topts,
```

File: tools/tpm2_startauthsession.c [hunk 10]
```diff
@@ -78,6 +84,14 @@ static bool on_option(char key, char *value) {
             return false;
         }
         break;
+    case 's':
```
Comment (idesai) [inline tools/tpm2_startauthsession.c:87]:
```suggestion
    case 'G':
```

File: tools/tpm2_startauthsession.c [hunk 11]
```diff
@@ -120,11 +134,12 @@ static bool tpm2_tool_onstart(tpm2_options **opts) {
         { "tpmkey-context", required_argument, NULL,  4 },
         { "hmac-session",   no_argument,       NULL,  5 },
         { "hash-algorithm", required_argument, NULL, 'g'},
+        { "sym-algorithm",  required_argument, NULL, 's'},
         { "session",        required_argument, NULL, 'S'},
         { "key-context",    required_argument, NULL, 'c'},
     };
 
-    *opts = tpm2_options_new("g:S:c:", ARRAY_LEN(topts), topts, on_option,
+    *opts = tpm2_options_new("g:s:S:c:", ARRAY_LEN(topts), topts, on_option,
```
Comment (idesai) [inline tools/tpm2_startauthsession.c:142]:
```suggestion
    *opts = tpm2_options_new("g:G:S:c:", ARRAY_LEN(topts), topts, on_option,
```

File: tools/tpm2_certify.c [hunk 12]
```diff
@@ -219,10 +178,10 @@ static tool_rc process_inputs(ESYS_CONTEXT *ectx) {
     /*
      * 3. Command specific initializations
      */
-    rc = set_scheme(ectx, ctx.signing_key.object.tr_handle, ctx.halg,
-        &ctx.scheme);
+    rc = tpm2_alg_util_get_signature_scheme(ectx,
+            ctx.signing_key.object.tr_handle, &ctx.halg, ctx.sig_scheme,
+            &ctx.scheme);
```
Comment (idesai) [inline tools/tpm2_certify.c:183]:
```suggestion
        ctx.signing_key.object.tr_handle, &ctx.halg, ctx.sig_scheme,
        &ctx.scheme);
```

**Review by idesai — CHANGES_REQUESTED** [review]

**Comment by idesai** [comment]
@mayuanchenma also can you please rebase to current master and resolve the conflicts. Thanks.

File: tools/tpm2_startauthsession.c [hunk 13]
```diff
@@ -120,11 +134,12 @@ static bool tpm2_tool_onstart(tpm2_options **opts) {
         { "tpmkey-context", required_argument, NULL,  4 },
         { "hmac-session",   no_argument,       NULL,  5 },
         { "hash-algorithm", required_argument, NULL, 'g'},
+        { "key-algorithm",  required_argument, NULL, 'G'},
         { "session",        required_argument, NULL, 'S'},
         { "key-context",    required_argument, NULL, 'c'},
     };
 
-    *opts = tpm2_options_new("g:S:c:", ARRAY_LEN(topts), topts, on_option,
+    *opts = tpm2_options_new("g:G:s:S:c:", ARRAY_LEN(topts), topts, on_option,
```
Comment (idesai) [inline tools/tpm2_startauthsession.c:142]:
```suggestion
    *opts = tpm2_options_new("g:G:S:c:", ARRAY_LEN(topts), topts, on_option,
```
```suggestion
    *opts = tpm2_options_new("g:G:s:S:c:", ARRAY_LEN(topts), topts, on_option,
```

Comment (idesai) [inline tools/tpm2_startauthsession.c:142]:
```suggestion
    *opts = tpm2_options_new("g:G:S:c:", ARRAY_LEN(topts), topts, on_option,
```
```suggestion
    *opts = tpm2_options_new("g:G:s:S:c:", ARRAY_LEN(topts), topts, on_option,
```

**Comment by idesai** [comment]
@mayuanchenma, this looks good. Can you please test the SM algorithms with this branch with openssl-3?

**Comment by mayuanchenma** [comment]
tpm2_createek -c /tmp/ek_ctx -G ecc_sm2 -t -u /tmp/pubkey -f PEM
test failed in openssl3
@idesai

**Comment by idesai** [comment]
> tpm2_createek -c /tmp/ek_ctx -G ecc_sm2 -t -u /tmp/pubkey -f PEM
> ERROR: Failed to create a EC public key: error:00000000:lib(0)::reason(0)
> test failed in openssl3
> @idesai

Any other debug information?

File: tools/tpm2_createek.c (hunk 4 above)
Comment (williamcroberts) [reply tools/tpm2_createek.c:205]: I belioeve its all high range NV indices

**Comment by mayuanchenma** [comment]
please merge the 2 commits. 
createek failed in openssl3 will be fixed afterward. first spent time studying openssl3.
@idesai

**Comment by williamcroberts** [comment]
> please merge the 2 commits. createek failed in openssl3 will be fixed afterward. first spent time studying openssl3. @idesai

We don't merge things that cause breaks to the CI, please upload the OpenSSL3 fix and we will merge assuming code review comments are corrected.

**Comment by mayuanchenma** [comment]
the  2 commits in openssl3 are ok.
please merge the 2 commits.
@idesai

**Comment by idesai** [comment]
> the 2 commits in openssl3 are ok.
> please merge the 2 commits.
> @idesai

@mayuanchenma thank you for submitting the pull request. Specifically the changes introduced in db06dd4d3 have to be addressed. Currently, we are able to compile the code with openssl3 but we cannot actually perform SM operations with it as the code is ifdef-compiled-out for openssl3.

**Comment by mayuanchenma** [comment]
> > the 2 commits in openssl3 are ok.
> > please merge the 2 commits.
> > @idesai
> 
> @mayuanchenma thank you for submitting the pull request. Specifically the changes introduced in [db06dd4](https://github.com/tpm2-software/tpm2-tools/commit/db06dd4d3c106abfb794f4f3738ee717f76ae52e) have to be addressed. Currently, we are able to compile the code with openssl3 but we cannot actually perform SM operations with it as the code is ifdef-compiled-out for openssl3.

SM operations test in openssl3 is ok. so ifdef-compiled-out for openssl3 is right.
@idesai

**Comment by idesai** [comment]
I was able to run this on a TPM with SM algs. So this works with a caveat that the checkquote needs to be updated too.

**Pull request merged**